circle/src/cfft.rs

```diff
@@ -0,0 +1,222 @@
+                    butterfly(lo_suffix, hi_suffix, t);
+                }
+            });
+        }
+        divide_by_height(&mut mat);
+        mat
+    }
+
+    pub fn icfft(&self, vec: Vec<F>) -> Vec<F> {
+        self.icfft_batch(RowMajorMatrix::new_col(vec)).values
+    }
+    pub fn icfft_batch<M: MatrixRowChunksMut<F>>(&self, mat: M) -> M {
+        let log_n = log2_strict_usize(mat.height());
+        self.coset_icfft_batch(mat, F::circle_two_adic_generator(log_n + 1))
+    }
+    /// The icfft: evaluating a polynomial in monomial basis over a domain
+    #[instrument(skip_all, fields(dims = %mat.dimensions()))]
+    pub fn coset_icfft_batch<M: MatrixRowChunksMut<F>>(&self, mat: M, shift: Complex<F>) -> M {
+        self.coset_icfft_batch_skipping_first_layers(mat, shift, 0)
+    }
+    #[instrument(skip_all, fields(dims = %mat.dimensions()))]
+    fn coset_icfft_batch_skipping_first_layers<M: MatrixRowChunksMut<F>>(
+        &self,
+        mut mat: M,
+        shift: Complex<F>,
+        num_skipped_layers: usize,
+    ) -> M {
+        let n = mat.height();
+        let log_n = log2_strict_usize(n);
+
+        let mut cache = self.0.borrow_mut();
+        let twiddles = cache.get_twiddles(log_n, shift, false);
+
+        for (i, twiddle) in twiddles.iter().rev().enumerate().skip(num_skipped_layers) {
+            let block_size = 1 << (i + 1);
+            let half_block_size = block_size >> 1;
+            assert_eq!(twiddle.len(), half_block_size);
+
+            mat.par_row_chunks_mut(block_size).for_each(|mut chunk| {
+                for (i, &t) in twiddle.iter().enumerate() {
+                    let (lo, hi) = chunk.row_pair_slices_mut(i, block_size - i - 1);
+                    let (lo_packed, lo_suffix) = F::Packing::pack_slice_with_suffix_mut(lo);
+                    let (hi_packed, hi_suffix) = F::Packing::pack_slice_with_suffix_mut(hi);
+                    ibutterfly(lo_packed, hi_packed, t.into());
+                    ibutterfly(lo_suffix, hi_suffix, t);
+                }
+            });
+        }
+
+        mat
+    }
+
+    #[instrument(skip_all, fields(dims = %mat.dimensions()))]
+    pub fn lde<M: MatrixRowChunksMut<F>>(
+        &self,
+        mut mat: M,
+        src_domain: CircleDomain<F>,
+        target_domain: CircleDomain<F>,
+    ) -> RowMajorMatrix<F> {
+        assert_eq!(mat.height(), src_domain.size());
+        assert!(target_domain.size() >= src_domain.size());
+        let added_bits = target_domain.log_n - src_domain.log_n;
+
+        // CFFT
+        mat = self.coset_cfft_batch(mat, src_domain.shift);
+
+        /*
+        To do an LDE, we could interleave zeros into the coefficients, but
+        the first `added_bits` layers will simply fill out the zeros with the
+        lower order values. (In `ibutterfly`, `hi` will start as zero, and
+        both `lo` and `hi` are set to `lo`). So instead, we do the tiling directly
+        and skip the first `added_bits` layers.
+        */
```

Comment:
Is this as efficient as doing the inverse FFT in the "coset-like" manner, that is to decompose the target domain into twin-cosets and apply the inverse FFT over each of these?